lib/DDG/Spice/PublicHolidays.pm

```diff
@@ -0,0 +1,121 @@
+package DDG::Spice::PublicHolidays;
+
+# ABSTRACT: Provides the list of public holidays in a calendar year for a country. 
+
+use strict;
+use DDG::Spice;
+use Locale::Country;
+
+spice is_cached => 1;
+spice proxy_cache_valid => "200 30d";
+spice wrap_jsonp_callback => 0;
+
+spice from => "(.+)/(.+)";
+spice to => 'https://api.xmltime.com/holidays?accesskey={{ENV{DDG_SPICE_TIME_AND_DATE_ACCESSKEY}}}&secretkey={{ENV{DDG_SPICE_TIME_AND_DATE_SECRETKEY}}}&version=2&callback={{callback}}&types=federal&country=$1&year=$2';
+
+my @triggers  = ('public holidays', 'national holidays', 'bank holidays', 'federal holidays');
+my $triggers  = join('|', @triggers);
+my $countries = join('|', share("countries.txt")->slurp(chomp => 1));
+
+triggers any => @triggers;
+    
+handle query_lc => sub {
+    my $query = $_;
+    
+    # Current year and users location are the defaults unless otherwise specified by the query
+    my $defaultYear = (localtime(time))[5] + 1900;
+    my $defaultCountry = $loc->country_name;
+    
+    # Regexes to match components of queries relevant to this IA
+    my $holidays = qr/$triggers/;
+    my $country  = qr/$countries/;
+    my $year     = qr/[1-9]{1}[0-9]{3}/;
+    
+    # Regexes to ignore optional words that can precede the year or country name
+    my $in       = qr/(?:in )?/;
+    my $inThe    = qr/(?:in the |in )?/;
+    
+    my $chosenYear;
+    my $chosenCountry;
+    
+    # Match: <trigger>
```

Comment:
This is good, but it's a little too verbose and I believe there's a simpler approach.

We already know one of the triggers exists in the query and we don't care to know which it is or where it is. So immediately we can strip it out to simplify our query. e.g. `$query =~ s/$triggers//;`

If there's a year in our query we can easy capture that with a simple regex `\d{4}` and then remove it from our query.

If we assume the queries only match our expected pattern and the query still has a word, we can check if it exists in our country hash. If the query has multiple remaining words we can check if any one of them is a country.

An approach like this should be simpler to implement and easier to maintain or improve